caom2pipe/client_composable.py

```diff
@@ -0,0 +1,665 @@
+        if checksum is not None and meta['md5sum'] != checksum:
+            logging.debug(
+                f'Different checksums: CADC {meta["md5sum"]} Source {checksum}'
+            )
+            retrieve = True
+        else:
+            logging.info(f'{f_name} already exists at CADC/{archive}')
+    except exceptions.NotFoundException:
+        retrieve = True
+
+    if retrieve:
+        logging.info(f'Retrieving {f_name} for {archive}')
+        fqn = os.path.join(working_dir, f_name)
+        mc.http_get(url, fqn)
+        mc.data_put(
+            cadc_client,
+            working_dir,
+            f_name,
+            archive,
+            stream,
+            mime_type,
+            mime_encoding=None,
+            metrics=metrics,
+        )
+
+
+def look_pull_and_put_v(
+    storage_name, f_name, working_dir, url, cadc_client, checksum, metrics
+):
+    """Checks to see if a file exists at CADC. If yes, stop. If no,
+    pull via https to local storage, then put to CADC storage.
+
+    TODO - stream
+    TODO - refactor to work with a single http session for archive.gemini.edu
+
+    :param storage_name The file name as it appears at CADC
+    :param f_name file name on disk for caching between the
+        pull and the put
+    :param working_dir together with f_name, location for caching
+    :param url for retrieving the file externally, if it does not exist
+    :param cadc_client access to the storage service
+    :param checksum what the CAOM observation says the checksum should be -
+        just the checksum part of ChecksumURI please, or the comparison will
+        always fail.
+    :param metrics track how long operations take
+    """
+    retrieve = False
+    try:
+        meta = mc._get_file_info(storage_name, cadc_client)
+        if checksum is not None and meta.md5sum != checksum:
+            logging.debug(
+                f'Different checksums: CADC {meta.md5sum} Source {checksum}'
+            )
+            retrieve = True
+        else:
+            logging.info(f'{f_name} already exists at CADC.')
+    except exceptions.NotFoundException:
+        retrieve = True
+
+    if retrieve:
+        logging.info(f'Retrieving {f_name}')
+        fqn = os.path.join(working_dir, f_name)
+        mc.http_get(url, fqn)
+        client_put(
+            cadc_client, working_dir, f_name, storage_name, metrics=metrics
+        )
+
+
+def query_tap_client(query_string, tap_client):
+    """
+    :param query_string ADQL
+    :param tap_client which client to query the service with
+    :returns an astropy votable instance."""
+
+    logging.debug(f'query_tap_client: execute query \n{query_string}')
+    buffer = StringIO()
+    tap_client.query(
+        query_string,
+        output_file=buffer,
+        data_only=True,
+        response_format='csv',
+    )
+    return Table.read(buffer.getvalue().split('\n'), format='csv')
+
+
+def repo_create(client, observation, metrics):
```

Comment:
When I originally wrote the functions, I did try to use a decorator. But the `Metrics` instance being used is long-lived, and I could not figure out how to work that instance into the decorator implementation. If you can point me to a way to do this, I think it would be much cleaner code.